hashing/hashing.cc

```diff
@@ -27,8 +28,50 @@ pair<ast::ParsedFile, core::UsageHash> rewriteAST(const core::GlobalState &origi
     return make_pair<ast::ParsedFile, core::UsageHash>(move(rewritten), subst.getAllNames());
 }
 
-unique_ptr<core::FileHash> computeFileHashForAST(unique_ptr<core::GlobalState> &lgs, core::UsageHash usageHash,
-                                                 ast::ParsedFile file) {
+bool isEmptyParseResult(const core::GlobalState &gs, const ast::ExpressionPtr &tree) {
+    if (ast::isa_tree<ast::EmptyTree>(tree)) {
+        return true;
+    }
+
+    auto *classDef = ast::cast_tree<ast::ClassDef>(tree);
+    if (classDef == nullptr || classDef->symbol != core::Symbols::root() || classDef->rhs.empty()) {
+        return false;
+    }
+
+    auto *rhsLiteral = ast::cast_tree<ast::Literal>(classDef->rhs[0]);
+    return rhsLiteral != nullptr && rhsLiteral->isNil(gs);
+}
+
+unique_ptr<core::FileHash> computeFileHashForAST(spdlog::logger &logger, unique_ptr<core::GlobalState> &lgs,
+                                                 core::UsageHash usageHash, ast::ParsedFile file) {
+    if (file.file.data(*lgs).hasParseErrors) {
+        if (isEmptyParseResult(*lgs, file.tree)) {
+            rapidjson::StringBuffer result;
+            rapidjson::Writer<rapidjson::StringBuffer> writer(result);
+
+            {
+                writer.StartObject();
+
+                writer.String("compute_file_hash_for_ast_empty");
+                writer.Bool(true);
+
+                writer.String("file_path");
+                writer.String(string(file.file.data(*lgs).path()));
+
+                writer.String("contents");
+                writer.String(string(file.file.data(*lgs).source()));
+
+                writer.EndObject();
+            }
+
+            logger.debug(result.GetString());
+
+            core::GlobalStateHash invalid;
+            invalid.hierarchyHash = core::GlobalStateHash::HASH_STATE_INVALID;
+            return make_unique<core::FileHash>(move(*lgs->hash()), move(usageHash));
```

Comment:
Did you mean to pass `invalid` in place of `*lgs->hash()` or something? (Otherwise I'm not sure what we're using `invalid` for.)

Reply:
good catch, i fixed that, and i wrote a test that would have caught this mistake.